synapse/handlers/space_summary.py

```diff
@@ -111,28 +116,75 @@ async def get_space_summary(
+
+                logger.debug(
+                    "Query of local room %s returned events %s",
+                    room_id,
+                    ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in events],
+                )
+
+                if room:
+                    rooms_result.append(room)
             else:
-                rooms, events = await self._summarize_remote_room(
+                fed_rooms, fed_events = await self._summarize_remote_room(
                     queue_entry,
                     suggested_only,
                     max_children,
                     exclude_rooms=processed_rooms,
                 )
 
-            logger.debug(
-                "Query of %s returned rooms %s, events %s",
-                queue_entry.room_id,
-                [room.get("room_id") for room in rooms],
-                ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in events],
-            )
-
-            rooms_result.extend(rooms)
-
-            # any rooms returned don't need visiting again
-            processed_rooms.update(cast(str, room.get("room_id")) for room in rooms)
+                # The results over federation might include rooms that the we,
+                # as the requesting server, are allowed to see, but the requesting
+                # user is not permitted see.
+                #
+                # A room may be visible if it is world readable, or if it is accessible
+                # to the specific user.
+                room_ids = set()
+                events = []
+                for room in fed_rooms:
+                    fed_room_id = room.get("room_id")
+                    if not fed_room_id:
+                        continue
+
+                    # Check whether the room is world readable from the response.
+                    include_room = room.get("world_readable") is True
+
+                    # If the room is not world-readable, get the allowed spaces from
+                    # the response.
+                    allowed_spaces = room.get("allowed_spaces")
+                    if not include_room and allowed_spaces:
+                        include_room = await self._event_auth_handler.is_user_in_rooms(
+                            allowed_spaces, requester
+                        )
+
+                    # Finally, check ourselves if we can access the room (this will
+                    # fail if we don't have any state for this room).
+                    if not include_room:
+                        include_room = await self._is_room_accessible(
+                            fed_room_id, requester, None
+                        )
+
+                    if include_room:
+                        rooms_result.append(room)
+                        room_ids.add(fed_room_id)
+
+                    # any rooms returned don't need visiting again (even if the user
+                    # didn't have access to them).
+                    processed_rooms.add(cast(str, room_id))
```

Comment:
Is this meant to be `fed_room_id`? If so, then we should probably explicitly check its a string rather than casting (and if not why do we need a cast as `room_id` is declared a `str`?).

Reply:
It is, yes! Good catch. 👍 

I agree, I'll add it to the check where we ensure it exists.